File: kdcreader_android/android/src/main/kotlin/koamtac/kdc/sdk/plugins/delegate/KNfcDelegate.kt
package koamtac.kdc.sdk.plugins.delegate

import android.content.Context
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import koamtac.kdc.sdk.KDCConstants
import koamtac.kdc.sdk.KDCReader
import koamtac.kdc.sdk.plugins.KConstants
import java.lang.ref.WeakReference

class KNfcDelegate(reader: KDCReader): KReaderDelegate() {
    companion object {
        private const val IS_NFC_INSTALLED = "isNFCInstalled"

        private const val ENABLE_NFC_POWER = "enableNFCPower"
        private const val IS_NFC_POWER_ENABLED = "isNFCPowerEnabled"

        private const val SET_NFC_DATA_FORMAT = "setNFCDataFormat"
        private const val GET_NFC_DATA_FORMAT = "getNFCDataFormat"

        private const val ENABLE_NFC_UID_ONLY = "enableNFCUIDOnly"
        private const val IS_NFC_UID_ONLY_ENABLED = "isNFCUIDOnlyEnabled"

        private const val ENABLE_NFC_EXTENDED_FORMAT = "enableNFCExtendedFormat"
        private const val IS_NFC_EXTENDED_FORMAT_ENABLED = "isNFCExtendedFormatEnabled"

        private val SUPPORTED_ACTIONS = HashSet(
            listOf(
                IS_NFC_INSTALLED,
                ENABLE_NFC_POWER,
                IS_NFC_POWER_ENABLED,
                SET_NFC_DATA_FORMAT,
                GET_NFC_DATA_FORMAT,
                ENABLE_NFC_UID_ONLY,
                IS_NFC_UID_ONLY_ENABLED,
                ENABLE_NFC_EXTENDED_FORMAT,
                IS_NFC_EXTENDED_FORMAT_ENABLED,
            )
        )

        private const val PARAM_NAME_ENABLE = "enable"
        private const val PARAM_NAME_FORMAT = "format"
    }

    init {
        wrReader = WeakReference(reader)
    }

    override fun isSupported(action: String): Boolean {
        return SUPPORTED_ACTIONS.contains(action)
    }

    override fun run(context: Context, call: MethodCall, result: MethodChannel.Result): Boolean {
        val reader = wrReader.get();

        if (reader == null) {
            sendErrorResult(KConstants.ErrorCode.instance.NULL, result)
            return true
        }

        when (call.method) {
            IS_NFC_INSTALLED -> isNFCInstalled(reader, call, result)

            ENABLE_NFC_POWER -> enableNFCPower(reader, call, result)
            IS_NFC_POWER_ENABLED -> isNFCPowerEnabled(reader, call, result)

            SET_NFC_DATA_FORMAT -> setNFCDataFormat(reader, call, result)
            GET_NFC_DATA_FORMAT -> getNFCDataFormat(reader, call, result)

            ENABLE_NFC_UID_ONLY -> enableNFCUIDOnly(reader, call, result)
            IS_NFC_UID_ONLY_ENABLED -> isNFCUIDOnlyEnabled(reader, call, result)

            ENABLE_NFC_EXTENDED_FORMAT -> enableNFCExtendedFormat(reader, call, result)
            IS_NFC_EXTENDED_FORMAT_ENABLED -> isNFCExtendedFormatEnabled(reader, call, result)
        }
        return true
    }

    private fun isNFCInstalled(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        val attached = reader.IsNFCInstalled()
        result.success(attached)
    }

    private fun enableNFCPower(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        val enable = call.argument<Boolean>(PARAM_NAME_ENABLE)

        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        enable?.let {
            val bRet = reader.EnableNFCPower(it)
            if (bRet) {
                result.success(true)
            } else {
                sendErrorResult(KConstants.ErrorCode.instance.FAILED, result)
            }
        } ?: sendErrorResult(KConstants.ErrorCode.instance.INVALID_PARAMETER, result)
    }

    private fun isNFCPowerEnabled(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        val enable = reader.IsNFCPowerEnabled()
        result.success(enable)
    }

    private fun setNFCDataFormat(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        val format = call.argument<Int>(PARAM_NAME_FORMAT)

        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        val dataFormat = format?.let {
            for (f in KDCConstants.NFCDataFormat.values()) {
                if (it == f.ordinal) {
                    return@let f
                }
            }
            null
        }

        dataFormat?.let {
            val bRet = reader.SetNFCDataFormat(it)

            if (bRet) {
                result.success(true)
            } else {
                sendErrorResult(KConstants.ErrorCode.instance.FAILED, result)
            }
        } ?: sendErrorResult(KConstants.ErrorCode.instance.INVALID_PARAMETER, result)
    }

    private fun getNFCDataFormat(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        val format = reader.GetNFCDataFormat()

        format?.let {
            result.success(it.ordinal)
        } ?: sendErrorResult(KConstants.ErrorCode.instance.FAILED, result)
    }

    private fun enableNFCUIDOnly(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        val enable = call.argument<Boolean>(PARAM_NAME_ENABLE)

        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        enable?.let {
            val bRet = reader.EnableNFCUIDOnly(it)

            if (bRet) {
                result.success(true)
            } else {
                sendErrorResult(KConstants.ErrorCode.instance.FAILED, result)
            }
        } ?: sendErrorResult(KConstants.ErrorCode.instance.INVALID_PARAMETER, result)
    }

    private fun isNFCUIDOnlyEnabled(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        val enable = reader.IsNFCUIDOnlyEnabled()
        result.success(enable)
    }

    private fun enableNFCExtendedFormat(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        val enable = call.argument<Boolean>(PARAM_NAME_ENABLE)

        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        enable?.let {
            val bRet = reader.EnableNFCExtendedFormat(it)

            if (bRet) {
                result.success(true)
            } else {
                sendErrorResult(KConstants.ErrorCode.instance.FAILED, result)
            }
        } ?: sendErrorResult(KConstants.ErrorCode.instance.INVALID_PARAMETER, result)
    }

    private fun isNFCExtendedFormatEnabled(reader: KDCReader, call: MethodCall, result: MethodChannel.Result) {
        if (!isConnected(reader)) {
            sendErrorResult(KConstants.ErrorCode.instance.NOT_CONNECTED, result)
            return
        }

        val enable = reader.IsNFCExtendedFormatEnabled()
        result.success(enable)
    }
}

File: kdcreader_ios/ios/Classes/delegate/KNfcDelegate.swift
//
//  KNfcDelegate.swift
//  kdcreader_ios
//
//  Created by koamtac on 3/21/24.
//

import Flutter

class KNfcDelegate: KReaderDelegate {
    private let IS_NFC_INSTALLED = "isNFCInstalled"

    private let ENABLE_NFC_POWER = "enableNFCPower"
    private let IS_NFC_POWER_ENABLED = "isNFCPowerEnabled"

    private let SET_NFC_DATA_FORMAT = "setNFCDataFormat"
    private let GET_NFC_DATA_FORMAT = "getNFCDataFormat"

    private let ENABLE_NFC_UID_ONLY = "enableNFCUIDOnly"
    private let IS_NFC_UID_ONLY_ENABLED = "isNFCUIDOnlyEnabled"

    private let ENABLE_NFC_EXTENDED_FORMAT = "enableNFCExtendedFormat"
    private let IS_NFC_EXTENDED_FORMAT_ENABLED = "isNFCExtendedFormatEnabled"

    /* PARAM */
    private let PARAM_NAME_ENABLE = "enable"
    private let PARAM_NAME_FORMAT = "format"
    
    override init(reader: KDCReader, converter: KConverter) {
        super.init(reader: reader, converter: converter)
        
        self.SUPPORTED_ACTIONS.append(contentsOf: [
            IS_NFC_INSTALLED,
            
            ENABLE_NFC_POWER,
            IS_NFC_POWER_ENABLED,
            
            SET_NFC_DATA_FORMAT,
            GET_NFC_DATA_FORMAT,
            
            ENABLE_NFC_UID_ONLY,
            IS_NFC_UID_ONLY_ENABLED,
            
            ENABLE_NFC_EXTENDED_FORMAT,
            IS_NFC_EXTENDED_FORMAT_ENABLED,
        ]);
    }
    
    override func run(_ call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        guard isConnected() else {
            sendErrorResult(KConstants.ErrorCode.NOT_CONNECTED, result:result)
            return true
        }
        
        switch call.method {
        case IS_NFC_INSTALLED:
            isNFCInstalled(call, result:result)

        case ENABLE_NFC_POWER:
            enableNFCPower(call, result:result)
            
        case IS_NFC_POWER_ENABLED:
            isNFCPowerEnabled(call, result:result)

        case SET_NFC_DATA_FORMAT:
            setNFCDataFormat(call, result:result)
            
        case GET_NFC_DATA_FORMAT:
            getNFCDataFormat(call, result:result)

        case ENABLE_NFC_UID_ONLY:
            enableNFCUIDOnly(call, result:result)
            
        case IS_NFC_UID_ONLY_ENABLED:
            isNFCUIDOnlyEnabled(call, result:result)

        case ENABLE_NFC_EXTENDED_FORMAT:
            enableNFCExtendedFormat(call, result:result)
            
        case IS_NFC_EXTENDED_FORMAT_ENABLED:
            isNFCExtendedFormatEnabled(call, result:result)
            
        default:
            break
        }
        return true
    }
    
    private func isNFCInstalled(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let attached = reader.isNFCInstalled()
        result(attached)
    }
    
    private func enableNFCPower(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? Dictionary<String, Any>,
              let enable = args[PARAM_NAME_ENABLE] as? Bool else {
            sendErrorResult(KConstants.ErrorCode.INVALID_PARAMETER, result:result)
            return
        }
        
        let bRet = reader.enableNFCPower(enable ? .ENABLE : .DISABLE)
        
        if bRet {
            result(true)
        } else {
            sendErrorResult(KConstants.ErrorCode.FAILED, result: result)
        }
    }

    private func isNFCPowerEnabled(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let enable = reader.isNFCPowerEnabled()
        result(enable == .ENABLE)
    }

    private func setNFCDataFormat(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? Dictionary<String, Any>,
              let v = args[PARAM_NAME_FORMAT] as? Int,
              let format = NFCDataFormat(rawValue: v) else {
            sendErrorResult(KConstants.ErrorCode.INVALID_PARAMETER, result:result)
            return
        }
        
        let bRet = reader.setNFCDataFormat(format)
        
        if bRet {
            result(true)
        } else {
            sendErrorResult(KConstants.ErrorCode.FAILED, result: result)
        }
    }

    private func getNFCDataFormat(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let format = reader.getNFCDataFormat().rawValue
        
        if let _ = NFCDataFormat(rawValue: format) {
            result(format)
        } else {
            sendErrorResult(KConstants.ErrorCode.FAILED, result: result)
        }
    }

    private func enableNFCUIDOnly(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? Dictionary<String, Any>,
              let enable = args[PARAM_NAME_ENABLE] as? Bool else {
            sendErrorResult(KConstants.ErrorCode.INVALID_PARAMETER, result:result)
            return
        }
        
        let bRet = reader.enableNFCUIDOnly(enable ? .ENABLE : .DISABLE)
        
        if bRet {
            result(true)
        } else {
            sendErrorResult(KConstants.ErrorCode.FAILED, result: result)
        }
    }

    private func isNFCUIDOnlyEnabled(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let enable = reader.isNFCUIDOnlyEnabled()
        result(enable == .ENABLE)
    }

    private func enableNFCExtendedFormat(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? Dictionary<String, Any>,
              let enable = args[PARAM_NAME_ENABLE] as? Bool else {
            sendErrorResult(KConstants.ErrorCode.INVALID_PARAMETER, result:result)
            return
        }
        
        let bRet = reader.enableNFCExtendedFormat(enable ? .ENABLE : .DISABLE)
        
        if bRet {
            result(true)
        } else {
            sendErrorResult(KConstants.ErrorCode.FAILED, result: result)
        }
    }

    private func isNFCExtendedFormatEnabled(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let enable = reader.isNFCExtendedFormatEnabled()
        result(enable == .ENABLE)
    }
}
